File: framework/src/main/java/nostalgia/framework/rom/DownloaderDelegate.java
package nostalgia.framework.rom;

import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.blankj.utilcode.util.Utils;
import com.liulishuo.okdownload.DownloadListener;
import com.liulishuo.okdownload.DownloadTask;
import com.liulishuo.okdownload.core.breakpoint.BreakpointInfo;
import com.liulishuo.okdownload.core.cause.EndCause;
import com.liulishuo.okdownload.core.cause.ResumeFailedCause;

import java.util.List;
import java.util.Map;

import nostalgia.framework.ui.gamegallery.adapter.GalleryPagerAdapter;
import nostalgia.framework.data.database.GameDescription;
import nostalgia.framework.utils.NLog;

public class DownloaderDelegate {

    public DownloaderDelegate(
            GalleryPagerAdapter.OnItemClickListener listener,
            ProgressBar progressBar,
            GameDescription gameDescription) {
        this.gameDescription = gameDescription;
        this.listener = listener;
        this.progressBar = progressBar;

        task = new DownloadTask.Builder(gameDescription.url, Utils.getApp().getFilesDir())
                .setFilename(gameDescription.name + ".nes")
                .setMinIntervalMillisCallbackProcess(30) // 下载进度回调的间隔时间（毫秒）
                .setPassIfAlreadyCompleted(false) // 任务过去已完成是否要重新下载
                .setPriority(10)
                .build();
    }

    public void startDownload() {
        task.enqueue(downloadListener);
    }

    private GameDescription gameDescription;
    private GalleryPagerAdapter.OnItemClickListener listener;
    private ProgressBar progressBar;
    DownloadTask task;

    DownloadListener downloadListener = new DownloadListener() {
        Long contentLength = 0L;
        Long curBytes = 0L;

        @Override
        public void taskStart(@NonNull DownloadTask task) {
            NLog.e("test111", "taskStart ==> ");
            progressBar.setProgress(0);
        }

        @Override
        public void connectTrialStart(@NonNull DownloadTask task, @NonNull Map<String, List<String>> requestHeaderFields) {

        }

        @Override
        public void connectTrialEnd(@NonNull DownloadTask task, int responseCode, @NonNull Map<String, List<String>> responseHeaderFields) {

        }

        @Override
        public void downloadFromBeginning(@NonNull DownloadTask task, @NonNull BreakpointInfo info, @NonNull ResumeFailedCause cause) {

        }

        @Override
        public void downloadFromBreakpoint(@NonNull DownloadTask task, @NonNull BreakpointInfo info) {

        }

        @Override
        public void connectStart(@NonNull DownloadTask task, int blockIndex, @NonNull Map<String, List<String>> requestHeaderFields) {

        }

        @Override
        public void connectEnd(@NonNull DownloadTask task, int blockIndex, int responseCode, @NonNull Map<String, List<String>> responseHeaderFields) {

        }

        @Override
        public void fetchStart(@NonNull DownloadTask task, int blockIndex, long contentLength) {
            this.contentLength = contentLength;
            curBytes = 0L;
            NLog.e("test111", "contentLength ==> $contentLength");
        }

        @Override
        public void fetchProgress(@NonNull DownloadTask task, int blockIndex, long increaseBytes) {
            curBytes += increaseBytes;
            int progress = (int) (curBytes * 100 / contentLength);
            NLog.e("test111", "progress ==> $progress");
            progressBar.setProgress(progress);
            gameDescription.path = task.getFile().getAbsolutePath();
        }

        @Override
        public void fetchEnd(@NonNull DownloadTask task, int blockIndex, long contentLength) {
            listener.onItemClick(gameDescription);
            NLog.e("test111", "fetchEnd ==> $blockIndex");
        }

        @Override
        public void taskEnd(@NonNull DownloadTask task, @NonNull EndCause cause, @Nullable Exception realCause) {
            NLog.e("test111", "EndCause ==> $cause");
            NLog.e("test111", "Exception ==> ", realCause);
        }
    };
}

File: gui/src/main/java/com/woohyman/gui/DownloaderDelegate.kt
package com.woohyman.gui

import android.view.View
import android.widget.ProgressBar
import com.blankj.utilcode.util.ToastUtils
import com.blankj.utilcode.util.Utils
import com.liulishuo.okdownload.DownloadListener
import com.liulishuo.okdownload.DownloadTask
import com.liulishuo.okdownload.core.breakpoint.BreakpointInfo
import com.liulishuo.okdownload.core.cause.EndCause
import com.liulishuo.okdownload.core.cause.ResumeFailedCause
import com.woohyman.gui.ui.gamegallery.adapter.GalleryPagerAdapter
import com.woohyman.keyboard.data.database.GameDescription
import com.woohyman.keyboard.utils.NLog

class DownloaderDelegate(
    private val listener: GalleryPagerAdapter.OnItemClickListener,
    private val progressBar: ProgressBar,
    private val gameDescription: GameDescription
) {
    fun startDownload() {
        task.enqueue(downloadListener)
    }

    var task: DownloadTask
    private var downloadListener: DownloadListener = object : DownloadListener {
        var contentLength = 0L
        var curBytes = 0L
        override fun taskStart(task: DownloadTask) {
            NLog.e("test111", "taskStart ==> ")
            progressBar.progress = 0
            progressBar.visibility = View.VISIBLE
        }

        override fun connectTrialStart(
            task: DownloadTask,
            requestHeaderFields: Map<String, List<String>>
        ) {
        }

        override fun connectTrialEnd(
            task: DownloadTask,
            responseCode: Int,
            responseHeaderFields: Map<String, List<String>>
        ) {
        }

        override fun downloadFromBeginning(
            task: DownloadTask,
            info: BreakpointInfo,
            cause: ResumeFailedCause
        ) {
        }

        override fun downloadFromBreakpoint(task: DownloadTask, info: BreakpointInfo) {}
        override fun connectStart(
            task: DownloadTask,
            blockIndex: Int,
            requestHeaderFields: Map<String, List<String>>
        ) {
        }

        override fun connectEnd(
            task: DownloadTask,
            blockIndex: Int,
            responseCode: Int,
            responseHeaderFields: Map<String, List<String>>
        ) {
        }

        override fun fetchStart(task: DownloadTask, blockIndex: Int, contentLength: Long) {
            this.contentLength = contentLength
            curBytes = 0L
            NLog.e("test111", "contentLength ==> \$contentLength")
        }

        override fun fetchProgress(task: DownloadTask, blockIndex: Int, increaseBytes: Long) {
            curBytes += increaseBytes
            val progress = (curBytes * 100 / contentLength).toInt()
            NLog.e("test111", "progress ==> \$progress")
            progressBar.progress = progress
            gameDescription.path = task.file!!.absolutePath
        }

        override fun fetchEnd(task: DownloadTask, blockIndex: Int, contentLength: Long) {
            listener.onItemClick(gameDescription)
            NLog.e("test111", "fetchEnd ==> \$blockIndex")
        }

        override fun taskEnd(task: DownloadTask, cause: EndCause, realCause: Exception?) {
            if (cause != EndCause.COMPLETED) {
                ToastUtils.showLong("错误信息 EndCause：" + cause + "，Exception：" + realCause?.message)
            }
            progressBar.visibility = View.INVISIBLE
        }
    }

    init {
        task = DownloadTask.Builder(gameDescription.url, Utils.getApp().filesDir)
            .setFilename(gameDescription.name + ".nes")
            .setMinIntervalMillisCallbackProcess(30) // 下载进度回调的间隔时间（毫秒）
            .setPassIfAlreadyCompleted(false) // 任务过去已完成是否要重新下载
            .setPriority(10)
            .build()
    }
}